batch/batch/front_end/validate.py

```diff
@@ -3,76 +3,138 @@
+        self.wrapped = wrapped
 
-# DEPRECATED:
-# command -> process/command
-# image -> process/image
-# mount_docker_socket -> process/mount_docker_socket
-# pvc_size -> resources/storage
-#
-# jobs_schema = [{
-#   'always_run': bool,
-#   'attributes': {str: str},
-#   'env': [{
-#     'name': str,
-#     'value': str
-#   }],
-#   'gcsfuse': [{"bucket": str, "mount_path": str, "read_only": bool}],
-#   'input_files': [{"from": str, "to": str}],
-#   'job_id': int,
-#   'output_files': [{"from": str, "to": str}],
-#   'parent_ids': [int],
-#   'port': int,
-#   'process': {
-#     'type': str,
-#     **process_schema[process_type]
-#   },
-#   'requester_pays_project': str,
-#   'network': str,
-#   'resources': {
-#     'memory': str,
-#     'cpu': str,
-#     'storage': str
-#   },
-#   'secrets': [{
-#     'namespace': str,
-#     'name': str,
-#     'mount_path': str
-#   }],
-#   'service_account': {
-#     'namespace': str,
-#     'name': str
-#   },
-#   'timeout': float or int
-# }]
-#
-# process_schema = {
-#     'docker': {
-#         'command': str,
-#         'image': str,
-#         'mount_docker_socket': str
-#     }
-# }
-
-JOB_KEYS = {'always_run', 'attributes', 'env', 'gcsfuse',
-            'input_files', 'job_id', 'mount_tokens',
-            'output_files', 'parent_ids', 'port', 'process',
-            'requester_pays_project', 'network', 'resources', 'secrets',
-            'service_account', 'timeout'}
-
-ENV_VAR_KEYS = {'name', 'value'}
-
-SECRET_KEYS = {'namespace', 'name', 'mount_path'}
-
-RESOURCES_KEYS = {'memory', 'cpu', 'storage'}
-
-FILE_KEYS = {'from', 'to'}
-
-GCSFUSE_KEYS = {'bucket', 'mount_path', 'read_only'}
-
-K8S_NAME_REGEXPAT = r'[a-z0-9](?:[-a-z0-9]*[a-z0-9])?(?:\.[a-z0-9](?:[-a-z0-9]*[a-z0-9])?)*'
-K8S_NAME_REGEX = re.compile(K8S_NAME_REGEXPAT)
+
+class TypedValidator:
+    def __init__(self, t):
+        self.t = t
+
+    def validate(self, name, obj):
+        if not isinstance(obj, self.t):
+            raise ValidationError(f'{name} is not {self.t}')
+
+
+class dictof(TypedValidator):
+    def __init__(self, vchecker):
+        super().__init__(dict)
+        self.vchecker = vchecker
+
+    def validate(self, name, obj):
+        super().validate(name, obj)
+        for k, v in obj.items():
+            if not isinstance(k, str):
+                raise ValidationError(f'{name} has non-str key')
+            self.vchecker.validate(f"{name}[{k}]", obj[k])
+
+
+class keyed(TypedValidator):
+    def __init__(self, keyed_checkers):
+        super().__init__(dict)
+        self.checkers = keyed_checkers
+
+    def __getitem__(self, key):
+        return self.checkers[key]
+
+    def validate(self, name, obj):
+        super().validate(name, obj)
+        for k in obj:
+            if k not in self.checkers:
+                raise ValidationError(f'unknown key in {name}: {k}')
+        for k, checker in self.checkers.items():
+            if isinstance(checker, required):
```

Comment:
The syntax looks a bit weird:
```python3
keyed({
    'foo': required(int),
    'bar': nullabe(int),
    'baz': int
})
```

What do you think of this instead:
```python3
keyed({
    required('foo'): int,
    'bar': nullabe(int),
    'baz': int
})
```
That puts the required ness notion on the right part (the key).

It also simplifies `validate` a bit:
```python3
    def validate(self, name, obj):
         super().validate(name, obj)
         for k in obj:
             if k not in self.checkers:
                 raise ValidationError(f'unknown key in {name}: {k}')
         for k, checker in self.checkers.items():
             if isinstance(k, required) and k not in obj:
                 raise ValidationError(f'{name} missing required key {k}.')
             checker.validate(f"{name}.{key}", obj[k])

```

Reply:
yeah, that looks good! I'll change